cmd/monitoring/monitoring.go

```diff
@@ -0,0 +1,179 @@
+)
+
+type monitor struct {
+	log    *logrus.Entry
+	pipcli azureclient.PublicIPAddressesClient
+
+	resourceGroup  string
+	subscriptionID string
+}
+
+type instance struct {
+	hostname string
+	b        *blackbox.Config
+}
+
+func (m *monitor) init(ctx context.Context, log *logrus.Entry) error {
+	authorizer, err := azureclient.NewAuthorizerFromEnvironment("")
+	if err != nil {
+		return err
+	}
+	if os.Getenv("RESOURCEGROUP") == "" {
+		return fmt.Errorf("RESOURCEGROUP environment variable must be set")
+	}
+	m.resourceGroup = os.Getenv("RESOURCEGROUP")
+	m.subscriptionID = os.Getenv("AZURE_SUBSCRIPTION_ID")
+	m.pipcli = azureclient.NewPublicIPAddressesClient(ctx, m.subscriptionID, authorizer)
+	m.log = log
+
+	return nil
+}
+
+func (m *monitor) listResourceGroupMonitoringHostnames(ctx context.Context, subscriptionID, resourceGroup string) (hostnames []string, err error) {
+	// get all external IP's used by VMSS
+	for {
+		hostnames = []string{}
+		for iter, err := m.pipcli.ListVirtualMachineScaleSetPublicIPAddressesComplete(ctx, resourceGroup, "ss-master"); iter.NotDone(); err = iter.Next() {
+			if err != nil {
+				m.log.Debug("waiting for url")
+				time.Sleep(5 * time.Second)
+			} else if iter.Value().IPAddress != nil {
+				hostnames = append(hostnames, *iter.Value().IPAddress)
+			}
+		}
+		if err == nil && len(hostnames) == 3 {
+			break
+		}
+	}
+	// get api server hostname
+	for {
+		ip, err := m.pipcli.Get(ctx, resourceGroup, "ip-apiserver", "")
+		if err != nil {
+			m.log.Debug("waiting for url")
+			time.Sleep(5 * time.Second)
+		} else if err == nil && ip.Location != nil {
+			hostnames = append(hostnames, fmt.Sprintf("%s.%s.cloudapp.azure.com", *ip.DNSSettings.DomainNameLabel, *ip.Location))
+			break
+		}
+	}
+	return hostnames, nil
+}
+
+func (m *monitor) run(ctx context.Context) error {
+	m.log.Info("fetching URLs\n")
+	hostnames, err := m.listResourceGroupMonitoringHostnames(ctx, m.subscriptionID, m.resourceGroup)
+	if err != nil {
+		return err
+	}
+
+	var instances []instance
+
+	for _, hostname := range hostnames {
+		m.log.Debugf("initiate blackbox monitor %s \n", hostname)
+		req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("https://%s/healthz", hostname), nil)
+		if err != nil {
+			return err
+		}
+		b := &blackbox.Config{
+			Cli: &http.Client{
+				Transport: &http.Transport{
+					TLSClientConfig: &tls.Config{
+						InsecureSkipVerify: true,
+					},
+					DisableKeepAlives: true,
+				},
+				Timeout: time.Second,
+			},
+			Req:              req,
+			Interval:         *interval,
+			LogInitialErrors: *logerrors,
+		}
+
+		instances = append(instances, struct {
+			hostname string
+			b        *blackbox.Config
+		}{
+			hostname: hostname,
+			b:        b,
+		})
+	}
+
+	for _, mon := range instances {
+		mon.b.Start(ctx)
+	}
+
+	ch := make(chan os.Signal)
+	signal.Notify(ch, os.Interrupt)
+	m.log.Info("collecting metrics... CTRL+C to stop\n")
+
+	for {
```

Comment:
This part is messy. So if we got stuck on L61, where we try to resolve hostnames before we create monitors this part of the code is not executed, so we never listen for `sigint` signal, so `kill -2 pid` does not work. Not sure how this can be solved now.